src/uct/tcp/tcp_listener.c

```diff
@@ -0,0 +1,216 @@
+    status = ucs_sys_fcntl_modfl(self->listener_fd, O_NONBLOCK, 0);
+    if (status != UCS_OK) {
+        status = UCS_ERR_IO_ERROR;
+        goto err_close_socket;
+    }
+
+    if (bind(self->listener_fd, saddr, socklen) < 0) {
+        ucs_error("bind(fd=%d addr=%s) failed: %m",
+                  self->listener_fd, ucs_sockaddr_str(saddr, ip_port_str,
+                                                      UCS_SOCKADDR_STRING_LEN));
+        status = (errno == EADDRINUSE) ? UCS_ERR_BUSY : UCS_ERR_IO_ERROR;
+        goto err_close_socket;
+    }
+
+    if (listen(self->listener_fd, backlog) < 0) {
+        ucs_error("listen(fd=%d addr=%s backlog=%d) failed: %m",
+                  self->listener_fd, ucs_sockaddr_str(saddr, ip_port_str,
+                                                      UCS_SOCKADDR_STRING_LEN),
+                  backlog);
+        status = UCS_ERR_IO_ERROR;
+        goto err_close_socket;
+    }
+
+    ucs_list_head_init(&self->fd_list);
+
+    async_ctx = tcp_sockcm->super.iface.worker->async;
+    status = ucs_async_set_event_handler(async_ctx->mode, self->listener_fd,
+                                         UCS_EVENT_SET_EVREAD |
+                                         UCS_EVENT_SET_EVERR,
+                                         uct_tcp_listener_conn_req_handler, self,
+                                         async_ctx);
+    if (status != UCS_OK) {
+        goto err_close_socket;
+    }
+
+    ucs_debug("created a TCP listener %p on cm %p with fd: %d. "
+              "listening on %s", self, cm, self->listener_fd,
+              ucs_sockaddr_str(saddr, ip_port_str, UCS_SOCKADDR_STRING_LEN));
+
+    return UCS_OK;
+
+err_close_socket:
+    close(self->listener_fd);
+err:
+    return status;
+}
+
+UCS_CLASS_CLEANUP_FUNC(uct_tcp_listener_t)
+{
+    uct_tcp_listener_ctx_t *listen_ctx, *tmp;
+    ucs_status_t status;
+
+    status = ucs_async_remove_handler(self->listener_fd, 1);
+    if (status != UCS_OK) {
+        ucs_warn("failed to remove event handler for fd %d: %s",
+                 self->listener_fd, ucs_status_string(status));
+    }
+
+    close(self->listener_fd);
+
+    UCS_ASYNC_BLOCK(self->super.cm->iface.worker->async);
+
+    ucs_list_for_each_safe(listen_ctx, tmp, &self->fd_list, list) {
+        ucs_async_remove_handler(listen_ctx->fd, 1);
+        close(listen_ctx->fd);
+        ucs_free(listen_ctx);
+    }
+
+    UCS_ASYNC_UNBLOCK(self->super.cm->iface.worker->async);
+}
+
+ucs_status_t uct_tcp_listener_reject(uct_listener_h listener,
+                                        uct_conn_request_h conn_request)
+{
+    return UCS_ERR_NOT_IMPLEMENTED;
+}
+
+ucs_status_t uct_tcp_listener_query(uct_listener_h listener,
+                                       uct_listener_attr_t *listener_attr)
+{
+    uct_tcp_listener_t *tcp_listener = ucs_derived_of(listener,
+                                                      uct_tcp_listener_t);
+    struct sockaddr_storage addr;
```

Comment:
seems like `struct sockaddr_storage` is not needed here, can use `struct sockaddr`

Reply:
sockaddr_storage is needed to have a correct sock_len (we need to support both ipv4 and ipv6).

Reply:
ok